### PR TITLE
Add support for stats collection in Hive for timestamp(p>3)

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -6262,6 +6262,65 @@ public void testAnalyzeUnpartitionedTable()
...
+        String catalog = getSession().getCatalog().get();
+        Session nanosecondsTimestamp = Session.builder(getSession())
+                // Disable column statistics collection when creating the table
+                .setCatalogSessionProperty(catalog, "collect_column_statistics_on_write", "false")
+                .setCatalogSessionProperty(catalog, "timestamp_precision", "NANOSECONDS")
+                .build();
+
+        Session microsecondsTimestamp = Session.builder(getSession())
+                .setCatalogSessionProperty(catalog, "timestamp_precision", "MICROSECONDS")
+                .build();
```
Comment (aalbu) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6277]:
```suggestion
        Session microsecondsTimestamp = withTimestampPrecision(getSession(), HiveTimestampPrecision.MICROSECONDS);
```

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 2]
```diff
@@ -6262,6 +6262,65 @@ public void testAnalyzeUnpartitionedTable()
...
+                        "TIMESTAMP '1988-04-08 02:03:04.112', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.113', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.111111', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.111112', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.111112', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.111113', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.111111111', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.111111112', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.111111112', " +
+                        "TIMESTAMP '1988-04-08 02:03:04.111111113' ", 12);
```
Comment (aalbu) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6296]: Can we have the last argument (12) on a different line?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 3]
```diff
@@ -6262,6 +6262,65 @@ public void testAnalyzeUnpartitionedTable()
...
+        assertUpdate(microsecondsTimestamp, "ANALYZE " + tableName, 12);
+        assertQuery("SHOW STATS FOR " + tableName,
+                "SELECT * FROM VALUES " +
+                        "('c_timestamp', null, 6.0, 0.0, null, null, null), " +
+                        "(null, null, null, null, 12.0, null, null)");
+
+        assertUpdate(millisecondsTimestamp, "ANALYZE " + tableName, 12);
+        assertQuery("SHOW STATS FOR " + tableName,
+                "SELECT * FROM VALUES " +
+                        "('c_timestamp', null, 3.0, 0.0, null, null, null), " +
```
Comment (aalbu) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6318]: Should we have a value that rounds up?

**Review by aalbu — COMMENTED** [review]
LGTM.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 1 above)
Comment (losipiuk) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6277]: thanks. I did not notice this helper.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 3 above)
Comment (losipiuk) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6318]: Added

**Comment by losipiuk** [comment]
ac

**Comment by losipiuk** [comment]
I added `Add testing for stats collection on write to Hive for timestamp(p>3)`

**Comment by losipiuk** [comment]
> please consider rebasing.

Yeah - I am alrady in the middle of that. Just spotted some test failure I need to fix.

**Comment by losipiuk** [comment]
Rebased on top of https://github.com/prestosql/presto/pull/5858

**Comment by losipiuk** [comment]
@findepi this should be good for review now.

File: presto-main/src/main/java/io/prestosql/operator/aggregation/DefaultApproximateCountDistinctAggregation.java [hunk 4]
```diff
@@ -75,7 +76,7 @@ public static void input(
     public static void input(
             @OperatorDependency(operator = XX_HASH_64, argumentTypes = "T") MethodHandle methodHandle,
             @AggregationState HyperLogLogState state,
-            @SqlType("T") Slice value)
+            @SqlType("T") Object value)
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/aggregation/DefaultApproximateCountDistinctAggregation.java:79]: Can you add a test within this commit?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java [hunk 5]
```diff
@@ -453,9 +454,13 @@ private static OptionalDouble getDoubleValue(ConnectorSession session, Type type
...
     }
 
-    private static OptionalLong getTimestampValue(Block block)
+    private static OptionalLong getTimestampValue(ConnectorSession session, Type type, Block block)
     {
-        return block.isNull(0) ? OptionalLong.empty() : OptionalLong.of(floorDiv(block.getLong(0, 0), MICROSECONDS_PER_SECOND));
+        if (block.isNull(0)) {
+            return OptionalLong.empty();
+        }
+        long epochMicros = ((SqlTimestamp) type.getObjectValue(session, block, 0)).getEpochMicros();
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java:462]: Do not use `. getObjectValue`, it's supposed to be used for sending output to the client.
Please use `.getLong` or `.getObject`, depending on `type.isShort`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java [hunk 6]
```diff
@@ -453,9 +454,13 @@ private static OptionalDouble getDoubleValue(ConnectorSession session, Type type
...
+        return OptionalLong.of(floorDiv(epochMicros, MICROSECONDS_PER_SECOND));
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java:463]: Here we truncate picos-of-micro, and then truncate micros-of-second.
This is correct for "min" value, but not for the "max"

Anyway, since the method is unused, please just remove it, not fix.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 7]
```diff
@@ -6336,6 +6336,64 @@ public void testAnalyzeUnpartitionedTable()
...
+    @Test
+    public void testAnalyzeTableTimestampWithPrecision()
+    {
+        String catalog = getSession().getCatalog().get();
+        Session nanosecondsTimestamp = Session.builder(withTimestampPrecision(getSession(), HiveTimestampPrecision.NANOSECONDS))
+                // Disable column statistics collection when creating the table
+                .setCatalogSessionProperty(catalog, "collect_column_statistics_on_write", "false")
+                .build();
+        Session microsecondsTimestamp = withTimestampPrecision(getSession(), HiveTimestampPrecision.MICROSECONDS);
+        Session millisecondsTimestamp = getSession();
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6382]: How do you  know millis is the default?
Maybe call `withTimestampPrecision` here too?
or verify

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 8]
```diff
@@ -6336,6 +6336,64 @@ public void testAnalyzeUnpartitionedTable()
...
+                            "('c_timestamp', null, null, null, null, null, null), " +
+                            "(null, null, null, null, 12.0, null, null)");
+
+            assertUpdate(nanosecondsTimestamp, "ANALYZE " + tableName, 12);
+            assertQuery("SHOW STATS FOR " + tableName,
+                    "SELECT * FROM VALUES " +
+                            "('c_timestamp', null, 9.0, 0.0, null, null, null), " +
+                            "(null, null, null, null, 12.0, null, null)");
+
+            assertUpdate(microsecondsTimestamp, "ANALYZE " + tableName, 12);
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6414]: It would be cleaner to drop the stats before running ANALYZE the second time, or create a new table

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 9]
```diff
@@ -6336,6 +6336,64 @@ public void testAnalyzeUnpartitionedTable()
...
+                // Disable column statistics collection when creating the table
+                .setCatalogSessionProperty(catalog, "collect_column_statistics_on_write", "false")
+                .build();
+        Session microsecondsTimestamp = withTimestampPrecision(getSession(), HiveTimestampPrecision.MICROSECONDS);
+        Session millisecondsTimestamp = getSession();
+
+        String tableName = "test_analyze_timestamp_with_precision";
+
+        try {
+            assertUpdate(nanosecondsTimestamp,
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6387]: move first arg to its own line too

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 10]
```diff
@@ -5750,6 +5750,40 @@ public void testCollectColumnStatisticsOnCreateTable()
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testCollectStatisticsOnCreateTableTimestampWithPrecision()
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:5754]: squash into prev commit?

**Review by findepi — APPROVED** [review]

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java (hunk 5 above)
Comment (losipiuk) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java:462]: I followed the pattern from method below.
```java
    private static Optional<BigDecimal> getDecimalValue(ConnectorSession session, Type type, Block block)
    {
        return block.isNull(0) ? Optional.empty() : Optional.of(((SqlDecimal) type.getObjectValue(session, block, 0)).toBigDecimal());
    }
```

What is the motivation to not use it. It makes the code simpler and the method has well defined semantics, right?

Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java:462]: Ok, so let me phrase this differently.
I think we generally avoid "the object representation" as I think it's supposed to be for the client procol.
Although i am aware javadoc doesn't say exactly this and we have some usages of the `getObjectValue` method outside of the communication impl.
So even, if it is the way I think it is, it is not documented.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 8 above)
Comment (losipiuk) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:6414]: I added dropping

**Comment by losipiuk** [comment]
AC

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java [hunk 11]
```diff
@@ -453,11 +454,6 @@ private static OptionalDouble getDoubleValue(ConnectorSession session, Type type
         return block.isNull(0) ? Optional.empty() : Optional.of(LocalDate.ofEpochDay(((SqlDate) type.getObjectValue(session, block, 0)).getDays()));
     }
 
-    private static OptionalLong getTimestampValue(Block block)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java:456]: Technically it would be cleaner to remove this in a separate commit

**Review by findepi — APPROVED** [review]

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java (hunk 11 above)
Comment (losipiuk) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java:456]: You are right. This should be done together with "drop min/max..."

**Comment by losipiuk** [comment]
@findepi @martint Would you consider this change user visible apart from https://github.com/prestosql/presto/issues/5678#issuecomment-724635858 ?
There can be potentially some performance improvement coming out of the fact that we have stats in place. Also `show stats` output would be different after this change. Enough to put entry in release notes?

**Comment by findepi** [comment]
well, as any other stats collection (to be consumed by the CBO), it's not user invisible, right?
i think it would be good to mention this in RNs.

**Comment by losipiuk** [comment]
> well, as any other stats collection (to be consumed by the CBO), it's not user invisible, right?
> i think it would be good to mention this in RNs.

I added https://github.com/prestosql/presto/issues/5678#issuecomment-724671666.
Not sure if not too techincal.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java (hunk 5 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/Statistics.java]: https://github.com/prestosql/presto/pull/5934
